{"cells":[{"cell_type":"markdown","id":"still-clarity","metadata":{"id":"still-clarity"},"source":["<h1>QIIME2 Workflow</h1>\n","\n","This notebook is a guide on working with QIIME2 with raw paired-end demultiplexed reads as the starting dataset. This notebook includes quality checking of raw reads, primer trimming, OTU picking, taxonomic assignment, and exporting data.\n","\n","This workflow was built with the following as the main references: <a href = 'https://github.com/LangilleLab/microbiome_helper/wiki/Amplicon-SOP-v2-(qiime2-2020.8)'>LangilleLab SOP</a>, <a href = 'https://docs.qiime2.org/2021.2/tutorials/moving-pictures/'>\"Moving pictures\" Tutorial</a>, and <a href = 'https://docs.qiime2.org/2021.2/tutorials/atacama-soils/'>\"Atacama soil microbiome\" tutorial</a>.\n","\n","Written for Day 1 of Bioinformatics Workshop by the Microbial Oceanography Laboratory. Credits: LBR dela Pe√±a, BW Hingpit, JB Quijano, D Purganan. "]},{"cell_type":"markdown","id":"nonprofit-editor","metadata":{"id":"nonprofit-editor"},"source":["<span style='color:red'>**Checkpoint:**</span> Import QIIME 2 python module to allow for inline visualizations"]},{"cell_type":"code","execution_count":null,"id":"contrary-party","metadata":{"id":"contrary-party"},"outputs":[],"source":["import qiime2 as q2"]},{"cell_type":"markdown","id":"paperback-summer","metadata":{"id":"paperback-summer"},"source":["### Making the manifest file\n","\n","Before we import our data, we have to make a **manifest file** that contains links to the forward and reverse file paths of each sample."]},{"cell_type":"code","execution_count":null,"id":"amateur-development","metadata":{"scrolled":true,"id":"amateur-development"},"outputs":[],"source":["import pandas as pd\n","import glob\n","import os\n","\n","sampleIDs, forwardpaths, reversepaths = [],[],[]\n","fpath = os.getcwd()+\"/raw-sequences/\"\n","for filepath in (glob.glob(fpath+\"*.gz\")):\n","    sample = filepath.split(\"/\")[-1].split(\"_\")[0]\n","    if sample not in sampleIDs:\n","        sampleIDs.append(sample)\n","    if \"R1_001.fastq.gz\" in filepath:\n","        forwardpaths.append(filepath)\n","    elif \"R2_001.fastq.gz\" in filepath:\n","        reversepaths.append(filepath)\n","\n","manifest =  pd.DataFrame({'sampleID': sorted(sampleIDs), 'forward-absolute-filepath': sorted(forwardpaths), 'reverse-absolute-filepath':sorted(reversepaths)} ) \n","with open('manifest.txt', 'w') as m:\n","    print(manifest.to_csv(sep='\\t', index=False, header=True), file=m)"]},{"cell_type":"markdown","id":"opening-accuracy","metadata":{"id":"opening-accuracy"},"source":["This [manifest file](data/manifest.txt) will show the sample ID and the absolute paths to the forward and reverse reads."]},{"cell_type":"markdown","id":"freelance-rebel","metadata":{"id":"freelance-rebel"},"source":["### Importing the sequences\n","Now that we prepared all the necessary files, we can make our first QIIME command: importing the sequence data."]},{"cell_type":"code","execution_count":null,"id":"signal-determination","metadata":{"id":"signal-determination"},"outputs":[],"source":["# Import the sequences\n","# Insert path to sequence folder after '--input-path'\n","!qiime tools import \\\n","    --type 'SampleData[PairedEndSequencesWithQuality]' \\\n","    --input-path manifest.txt \\\n","    --output-path raw-sequences/seqs.qza \\\n","    --input-format PairedEndFastqManifestPhred33V2"]},{"cell_type":"markdown","id":"champion-louisville","metadata":{"id":"champion-louisville"},"source":["This converts the sequence data into a **QIIME artifact**. Artifacts have the extension '.qza'"]},{"cell_type":"markdown","id":"thrown-substitute","metadata":{"id":"thrown-substitute"},"source":["### Quality checking\n","Our sequences are already *demultiplexed*, meaning they are already separated into different samples. We can use the `demux` plugin instead to visualize our sequences. **QIIME visualizations** have the extension '.qzv'. The .qzv files can be viewed in  http://view.qiime2.org or we can import the `qiime2` module to view the visualizations inline."]},{"cell_type":"code","execution_count":null,"id":"monetary-strand","metadata":{"scrolled":true,"id":"monetary-strand"},"outputs":[],"source":["# Make summary of the QIIME2 artifact (.qza file)\n","!qiime demux summarize \\\n","    --i-data  raw-sequences/seqs.qza \\\n","    --o-visualization raw-sequences/seqs.qzv\n","\n","# Visualize\n","q2.Visualization.load('raw-sequences/seqs.qzv')"]},{"cell_type":"markdown","id":"measured-twist","metadata":{"id":"measured-twist"},"source":["Open the visualization summary and go to the **Interactive Quality Plot**. Here, we can see the average quality score of the reads at each position. In general, we want to maintain a score above 30. "]},{"cell_type":"markdown","id":"rational-conservative","metadata":{"id":"rational-conservative"},"source":["## Data Processing\n","To prepare our sequences, we have to perform several steps:\n","1. Trim primers\n","2. DADA2:\n","- Merge paired-end reads\n","- Filter sequences by quality\n","- Dereplicate"]},{"cell_type":"markdown","id":"dimensional-batman","metadata":{"id":"dimensional-batman"},"source":["### Trim primers\n","To remove the primers in our sequences, we use the `cutadapt` plugin. The primers used were E572F/E1009R, which have **18bp** and **20bp** lengths, respectively. Removing the primers is important especially if there are ambiguous bases, which might get confused as chimeric or low quality positions. You can explore more about the primer sequences, length, and predicted amplicon size in this excellent app [PR-2 Primers](https://app.pr2-primers.org/). "]},{"cell_type":"code","execution_count":null,"id":"military-republic","metadata":{"id":"military-republic"},"outputs":[],"source":["# Make directory\n","%mkdir cleanup \\\n","\n","!qiime cutadapt trim-paired \\\n","    --i-demultiplexed-sequences raw-sequences/seqs.qza \\\n","    --p-front-f CYGCGGTAATTCCAGCTC \\\n","    --p-front-r AYGGTATCTRATCRTCTTYG \\\n","    --p-error-rate 0 \\\n","    --o-trimmed-sequences cleanup/primer-trimmed-seqs.qza"]},{"cell_type":"code","execution_count":null,"id":"egyptian-group","metadata":{"id":"egyptian-group"},"outputs":[],"source":["# Make summary of the QIIME2 artifact (.qza file)\n","# To check whether the primers were removed\n","!qiime demux summarize \\\n","    --i-data  cleanup/primer-trimmed-seqs.qza \\\n","    --o-visualization cleanup/primer-trimmed-seqs.qzv \\\n","\n","# Visualize\n","q2.Visualization.load('cleanup/primer-trimmed-seqs.qzv')"]},{"cell_type":"code","execution_count":null,"id":"criminal-drill","metadata":{"id":"criminal-drill","outputId":"725175ca-65b2-4cd9-bbb5-3ddea5ad40e2"},"outputs":[{"data":{"text/html":["<div><img onload=\"(function(div, url){\n","if (typeof require !== 'undefined') {\n","    var baseURL = require.toUrl('').split('/').slice(0, -2).join('/');\n","} else {\n","    var baseURL = JSON.parse(\n","        document.getElementById('jupyter-config-data').innerHTML\n","    ).baseUrl.slice(0, -1);\n","}\n","url = baseURL + url;\n","fetch(url).then(function(res) {\n","    if (res.status === 404) {\n","        div.innerHTML = 'Install QIIME 2 Jupyter extension with:<br />' +\n","                        '<code>jupyter serverextension enable --py qiime2' +\n","                        ' --sys-prefix</code><br />then restart your server.' +\n","                        '<br /><br />(Interactive output not available on ' +\n","                        'static notebook viewer services like nbviewer.)';\n","    } else if (res.status === 409) {\n","        div.innerHTML = 'Visualization no longer in scope. Re-run this cell' +\n","                        ' to see the visualization.';\n","    } else if (res.ok) {\n","        url = res.url;\n","        div.innerHTML = '<iframe src=\\'' + url + '\\' style=\\'' +\n","                        'width: 100%; height: 700px; border: 0;\\'>' +\n","                        '</iframe><hr />Open in a: <a href=\\'' + url + '\\'' +\n","                        ' target=\\'_blank\\'>new window</a>'\n","    } else {\n","        div.innerHTML = 'Something has gone wrong. Check notebook server for' +\n","                        ' errors.';\n","    }\n","});\n","})(this.parentElement, '/qiime2/redirect?location=/tmp/qiime2-archive-g7z8cpir')\" src=\"data:image/gif;base64,R0lGODlhAQABAIAAAP///wAAACH5BAEAAAAALAAAAAABAAEAAAICRAEAOw==\" /></div>"],"text/plain":["<visualization: Visualization uuid: 14bb287e-f284-4d16-8146-ad4d97be9622>"]},"execution_count":7,"metadata":{},"output_type":"execute_result"}],"source":["q2.Visualization.load('cleanup/primer-trimmed-seqs.qzv')"]},{"cell_type":"markdown","id":"musical-liability","metadata":{"id":"musical-liability"},"source":["### DADA2\n","DADA2 is a pipeline for detecting and correcting (where possible) Illumina amplicon sequence data. As implemented in the q2-dada2 plugin, this quality control process will additionally filter any phiX reads (commonly present in marker gene Illumina sequence data) that are identified in the sequencing data, and will filter chimeric sequences. DADA2 includes quality filtering, dereplicating, chimera filtering and singleton filtering.\n","\n","<span style='color:red'>**Note:**</span> Several trimming and truncating positions will be done to check the number of sequences that will pass DADA2 filter"]},{"cell_type":"markdown","id":"infrared-miami","metadata":{"id":"infrared-miami"},"source":["In DADA2, paired-end sequences are recommended to have at least 12-20 nt overlap. Considering that the 18S rRNA gene has an average length of 350-450 bases, trimming and truncation of paired reads should allow the recommended overlap."]},{"cell_type":"markdown","id":"satisfactory-victim","metadata":{"id":"satisfactory-victim"},"source":["<span style='color:blue'>Un-trimmed and un-truncated reads</span>"]},{"cell_type":"code","execution_count":null,"id":"coordinated-consent","metadata":{"id":"coordinated-consent"},"outputs":[],"source":["# Un-trimmed and un-truncated\n","!qiime dada2 denoise-paired \\\n","    --i-demultiplexed-seqs cleanup/primer-trimmed-seqs.qza \\\n","    --p-trunc-len-f 0 \\\n","    --p-trunc-len-r 0 \\\n","    --o-table cleanup/table-all.qza \\\n","    --o-representative-sequences cleanup/rep-seqs-all.qza \\\n","    --o-denoising-stats cleanup/denoising-stats-all.qza"]},{"cell_type":"code","execution_count":null,"id":"gross-economics","metadata":{"scrolled":true,"id":"gross-economics"},"outputs":[],"source":["# View statistics\n","!qiime metadata tabulate \\\n","    --m-input-file cleanup/denoising-stats-all.qza \\\n","    --o-visualization cleanup/denoising-stats-all.qzv \\\n","\n","q2.Visualization.load('cleanup/denoising-stats-all.qzv')"]},{"cell_type":"markdown","id":"bigger-courtesy","metadata":{"id":"bigger-courtesy"},"source":["<span style='color:blue'>Truncated at the 250th nucleotide</span>"]},{"cell_type":"code","execution_count":null,"id":"sustainable-elephant","metadata":{"id":"sustainable-elephant"},"outputs":[],"source":["# Truncated at 250th nucleotide\n","!qiime dada2 denoise-paired \\\n","    --i-demultiplexed-seqs cleanup/primer-trimmed-seqs.qza \\\n","    --p-trunc-len-f 250 \\\n","    --p-trunc-len-r 250 \\\n","    --o-table cleanup/table-250.qza \\\n","    --o-representative-sequences cleanup/rep-seqs-250.qza \\\n","    --o-denoising-stats cleanup/denoising-stats-250.qza"]},{"cell_type":"code","execution_count":null,"id":"formed-issue","metadata":{"id":"formed-issue"},"outputs":[],"source":["# View statistics\n","!qiime metadata tabulate \\\n","    --m-input-file cleanup/denoising-stats-250.qza \\\n","    --o-visualization cleanup/denoising-stats-250.qzv \\\n","\n","q2.Visualization.load('cleanup/denoising-stats-250.qzv')"]},{"cell_type":"code","execution_count":null,"id":"allied-button","metadata":{"id":"allied-button","outputId":"ec98b4e3-2f04-4f1b-d102-c758453f5118"},"outputs":[{"data":{"text/html":["<div><img onload=\"(function(div, url){\n","if (typeof require !== 'undefined') {\n","    var baseURL = require.toUrl('').split('/').slice(0, -2).join('/');\n","} else {\n","    var baseURL = JSON.parse(\n","        document.getElementById('jupyter-config-data').innerHTML\n","    ).baseUrl.slice(0, -1);\n","}\n","url = baseURL + url;\n","fetch(url).then(function(res) {\n","    if (res.status === 404) {\n","        div.innerHTML = 'Install QIIME 2 Jupyter extension with:<br />' +\n","                        '<code>jupyter serverextension enable --py qiime2' +\n","                        ' --sys-prefix</code><br />then restart your server.' +\n","                        '<br /><br />(Interactive output not available on ' +\n","                        'static notebook viewer services like nbviewer.)';\n","    } else if (res.status === 409) {\n","        div.innerHTML = 'Visualization no longer in scope. Re-run this cell' +\n","                        ' to see the visualization.';\n","    } else if (res.ok) {\n","        url = res.url;\n","        div.innerHTML = '<iframe src=\\'' + url + '\\' style=\\'' +\n","                        'width: 100%; height: 700px; border: 0;\\'>' +\n","                        '</iframe><hr />Open in a: <a href=\\'' + url + '\\'' +\n","                        ' target=\\'_blank\\'>new window</a>'\n","    } else {\n","        div.innerHTML = 'Something has gone wrong. Check notebook server for' +\n","                        ' errors.';\n","    }\n","});\n","})(this.parentElement, '/qiime2/redirect?location=/tmp/qiime2-archive-ks85rtty')\" src=\"data:image/gif;base64,R0lGODlhAQABAIAAAP///wAAACH5BAEAAAAALAAAAAABAAEAAAICRAEAOw==\" /></div>"],"text/plain":["<visualization: Visualization uuid: acdcf06b-b9ba-4fa9-980c-70e670a6bfe5>"]},"execution_count":3,"metadata":{},"output_type":"execute_result"}],"source":["q2.Visualization.load('cleanup/denoising-stats-250.qzv')"]},{"cell_type":"markdown","id":"prescribed-inflation","metadata":{"id":"prescribed-inflation"},"source":["## Assigning taxonomy\n","After quality filtering, the resulting data can be explored using `feature-table summarize` and `feature table tabulate-seqs`. The former command will give information on how many sequences are associated with each sample and with each feature (ASVs), histograms of those and some related summary statistics while the latter will provide a mapping of feature IDs to sequences, and provide links to easily BLAST each sequence against the NCBI nt database."]},{"cell_type":"code","execution_count":null,"id":"fancy-jewelry","metadata":{"id":"fancy-jewelry"},"outputs":[],"source":["# Summarize ASV table (un-truncated/un-trimmed)\n","!qiime feature-table summarize \\\n","    --i-table cleanup/table-all.qza \\\n","    --o-visualization cleanup/table-all.qzv \\\n","    --m-sample-metadata-file metadata.tsv\n","\n","# Visualize\n","q2.Visualization.load('cleanup/table-all.qzv')"]},{"cell_type":"code","execution_count":null,"id":"continuous-highlight","metadata":{"id":"continuous-highlight"},"outputs":[],"source":["# Summarize ASV table (truncated at 250th nucleotide)\n","!qiime feature-table summarize \\\n","    --i-table cleanup/table-250.qza \\\n","    --o-visualization cleanup/table-250.qzv \\\n","    --m-sample-metadata-file metadata.tsv\n","\n","# Visualize\n","q2.Visualization.load('cleanup/table-250.qzv')"]},{"cell_type":"code","execution_count":null,"id":"accurate-hopkins","metadata":{"id":"accurate-hopkins","outputId":"fffb5589-81c7-4d03-e3ef-31c1ef816ce0"},"outputs":[{"data":{"text/html":["<div><img onload=\"(function(div, url){\n","if (typeof require !== 'undefined') {\n","    var baseURL = require.toUrl('').split('/').slice(0, -2).join('/');\n","} else {\n","    var baseURL = JSON.parse(\n","        document.getElementById('jupyter-config-data').innerHTML\n","    ).baseUrl.slice(0, -1);\n","}\n","url = baseURL + url;\n","fetch(url).then(function(res) {\n","    if (res.status === 404) {\n","        div.innerHTML = 'Install QIIME 2 Jupyter extension with:<br />' +\n","                        '<code>jupyter serverextension enable --py qiime2' +\n","                        ' --sys-prefix</code><br />then restart your server.' +\n","                        '<br /><br />(Interactive output not available on ' +\n","                        'static notebook viewer services like nbviewer.)';\n","    } else if (res.status === 409) {\n","        div.innerHTML = 'Visualization no longer in scope. Re-run this cell' +\n","                        ' to see the visualization.';\n","    } else if (res.ok) {\n","        url = res.url;\n","        div.innerHTML = '<iframe src=\\'' + url + '\\' style=\\'' +\n","                        'width: 100%; height: 700px; border: 0;\\'>' +\n","                        '</iframe><hr />Open in a: <a href=\\'' + url + '\\'' +\n","                        ' target=\\'_blank\\'>new window</a>'\n","    } else {\n","        div.innerHTML = 'Something has gone wrong. Check notebook server for' +\n","                        ' errors.';\n","    }\n","});\n","})(this.parentElement, '/qiime2/redirect?location=/tmp/qiime2-archive-eieydpvy')\" src=\"data:image/gif;base64,R0lGODlhAQABAIAAAP///wAAACH5BAEAAAAALAAAAAABAAEAAAICRAEAOw==\" /></div>"],"text/plain":["<visualization: Visualization uuid: dbd97667-ce5d-4409-b882-469ab8ff1d08>"]},"execution_count":2,"metadata":{},"output_type":"execute_result"}],"source":["# Visualize\n","q2.Visualization.load('cleanup/table-250.qzv')"]},{"cell_type":"markdown","id":"introductory-stadium","metadata":{"id":"introductory-stadium"},"source":["The results show how many OTUs were determined in our samples in the overview. In the *Interactive Sample Detail* tab, you can see how many OTUs were detected in each sample. In the last tab (*Feature Detail*), you can see the OTU ID, their frequencies, and occurence in the samples. \n","üßê What is the most frequently detected OTU?\n","\n","At this point, sequences truncated at the 250th nucleotide base will be used as more sequences were retrieved."]},{"cell_type":"code","execution_count":null,"id":"violent-antibody","metadata":{"id":"violent-antibody"},"outputs":[],"source":["# Map OTUs to sequences\n","!qiime feature-table tabulate-seqs \\\n","    --i-data cleanup/rep-seqs-250.qza \\\n","    --o-visualization cleanup/rep-seqs-250.qzv\n","\n","# Visualize\n","q2.Visualization.load('cleanup/rep-seqs-250.qzv')"]},{"cell_type":"markdown","id":"celtic-generation","metadata":{"id":"celtic-generation"},"source":["### Taxonomy assignment using SILVA\n","\n","To annotate the metabarcoding data, we use a reference database which will classify the sequences to their taxonomic identities using the plugin `sci-kit learn`. For 18S rRNA eukaryotic data, we will use a classifier made from SILVA 138.1 optimized for the specific region targeted by the primers used in this run (18S V4 region). Other references for eukaryotic sequences can be used, such as the [PR2 database](https://pr2-database.org/) which has high-quality reference sequences curated by other experts."]},{"cell_type":"code","execution_count":null,"id":"dangerous-trail","metadata":{"id":"dangerous-trail"},"outputs":[],"source":["# Classify using sci-kit learn (sklearn)\n","!qiime feature-classifier classify-sklearn \\\n","    --i-classifier ../refdb-2/silva-138.1-euk-v4-200700-classifier.qza \\\n","    --i-reads cleanup/rep-seqs-250.qza \\\n","    --o-classification cleanup/asv-taxa-250.qza"]},{"cell_type":"code","execution_count":null,"id":"persistent-treatment","metadata":{"id":"persistent-treatment"},"outputs":[],"source":["# Visualize\n","!qiime metadata tabulate \\\n","    --m-input-file cleanup/asv-taxa-250.qza \\\n","    --o-visualization cleanup/asv-taxa-250.qzv\n","\n","q2.Visualization.load('cleanup/asv-taxa-250.qzv')"]},{"cell_type":"markdown","id":"amateur-litigation","metadata":{"id":"amateur-litigation"},"source":["We can view interactive taxonomic barplot to see the composition of each sample.\n","\n","After loading the visualization, select *Level* to 7 or 8 (when Kingdom is included) to view at the most resolved classification."]},{"cell_type":"code","execution_count":null,"id":"beneficial-packet","metadata":{"id":"beneficial-packet"},"outputs":[],"source":["# sklearn\n","!qiime taxa barplot \\\n","    --i-table cleanup/table-250.qza \\\n","    --i-taxonomy cleanup/asv-taxa-250.qza \\\n","    --m-metadata-file metadata.tsv \\\n","    --o-visualization cleanup/bar-plots-asv-250.qzv\n","\n","q2.Visualization.load('cleanup/bar-plots-asv-250.qzv')"]},{"cell_type":"markdown","id":"proud-athletics","metadata":{"id":"proud-athletics"},"source":["## Filtering ASV table and representative sequences\n","The ASV table and representative sequences should be filtered to exclude putative metazoans (animal, multi-cellular chlorophytes and fungal sequencces). This filtering is done to focus on protists only."]},{"cell_type":"code","execution_count":null,"id":"addressed-regulation","metadata":{"id":"addressed-regulation"},"outputs":[],"source":["# Filtering ASV table\n","!qiime taxa filter-table \\\n","    --i-table cleanup/table-250.qza \\\n","    --i-taxonomy cleanup/asv-taxa-250.qza \\\n","    --p-exclude k__Rhodophyceae,k__Fungi,k__Animalia,c__Embryophyta,c__Ulvophyceae \\\n","    --o-filtered-table cleanup/asv-table-nometaz.qza \\\n","\n","# Filtering representative sequences\n","!qiime taxa filter-seqs \\\n","    --i-sequences cleanup/rep-seqs-250.qza \\\n","    --i-taxonomy cleanup/asv-taxa-250.qza \\\n","    --p-exclude  k__Rhodophyceae,k__Fungi,k__Animalia,c__Embryophyta,c__Ulvophyceae \\\n","    --o-filtered-sequences cleanup/asv-rep-seqs-nometaz-250.qza"]},{"cell_type":"markdown","id":"dying-vision","metadata":{"id":"dying-vision"},"source":["Double check taxonomy barplots"]},{"cell_type":"code","execution_count":null,"id":"unlikely-president","metadata":{"id":"unlikely-president"},"outputs":[],"source":["# sklearn\n","!qiime taxa barplot \\\n","    --i-table cleanup/asv-table-nometaz.qza \\\n","    --i-taxonomy cleanup/asv-taxa-250.qza \\\n","    --m-metadata-file metadata.tsv \\\n","    --o-visualization cleanup/bar-plots-asv-nometaz.qzv"]},{"cell_type":"code","execution_count":null,"id":"alpha-violin","metadata":{"scrolled":true,"id":"alpha-violin","outputId":"9aa27265-9b87-4a2f-8964-1a6edef46af8"},"outputs":[{"data":{"text/html":["<div><img onload=\"(function(div, url){\n","if (typeof require !== 'undefined') {\n","    var baseURL = require.toUrl('').split('/').slice(0, -2).join('/');\n","} else {\n","    var baseURL = JSON.parse(\n","        document.getElementById('jupyter-config-data').innerHTML\n","    ).baseUrl.slice(0, -1);\n","}\n","url = baseURL + url;\n","fetch(url).then(function(res) {\n","    if (res.status === 404) {\n","        div.innerHTML = 'Install QIIME 2 Jupyter extension with:<br />' +\n","                        '<code>jupyter serverextension enable --py qiime2' +\n","                        ' --sys-prefix</code><br />then restart your server.' +\n","                        '<br /><br />(Interactive output not available on ' +\n","                        'static notebook viewer services like nbviewer.)';\n","    } else if (res.status === 409) {\n","        div.innerHTML = 'Visualization no longer in scope. Re-run this cell' +\n","                        ' to see the visualization.';\n","    } else if (res.ok) {\n","        url = res.url;\n","        div.innerHTML = '<iframe src=\\'' + url + '\\' style=\\'' +\n","                        'width: 100%; height: 700px; border: 0;\\'>' +\n","                        '</iframe><hr />Open in a: <a href=\\'' + url + '\\'' +\n","                        ' target=\\'_blank\\'>new window</a>'\n","    } else {\n","        div.innerHTML = 'Something has gone wrong. Check notebook server for' +\n","                        ' errors.';\n","    }\n","});\n","})(this.parentElement, '/qiime2/redirect?location=/tmp/qiime2-archive-y8a4e98k')\" src=\"data:image/gif;base64,R0lGODlhAQABAIAAAP///wAAACH5BAEAAAAALAAAAAABAAEAAAICRAEAOw==\" /></div>"],"text/plain":["<visualization: Visualization uuid: a28df0a7-f5c9-4ff1-8222-6fec7fbdceaf>"]},"execution_count":5,"metadata":{},"output_type":"execute_result"}],"source":["# Visualize\n","q2.Visualization.load('cleanup/bar-plots-asv-nometaz.qzv')"]},{"cell_type":"markdown","id":"acoustic-winter","metadata":{"id":"acoustic-winter"},"source":["Double check ASV table"]},{"cell_type":"code","execution_count":null,"id":"united-retreat","metadata":{"scrolled":true,"id":"united-retreat"},"outputs":[],"source":["# Summarize ASV table (truncated at 250th nucleotide)\n","!qiime feature-table summarize \\\n","    --i-table cleanup/asv-table-nometaz.qza \\\n","    --o-visualization cleanup/asv-table-nometaz.qzv \\\n","    --m-sample-metadata-file metadata.tsv\n","\n","# Visualize\n","q2.Visualization.load('cleanup/asv-table-nometaz.qzv')"]},{"cell_type":"code","execution_count":null,"id":"blind-authorization","metadata":{"id":"blind-authorization","outputId":"927d058e-c6a7-42a8-fac5-e5328fc97c6f"},"outputs":[{"data":{"text/html":["<div><img onload=\"(function(div, url){\n","if (typeof require !== 'undefined') {\n","    var baseURL = require.toUrl('').split('/').slice(0, -2).join('/');\n","} else {\n","    var baseURL = JSON.parse(\n","        document.getElementById('jupyter-config-data').innerHTML\n","    ).baseUrl.slice(0, -1);\n","}\n","url = baseURL + url;\n","fetch(url).then(function(res) {\n","    if (res.status === 404) {\n","        div.innerHTML = 'Install QIIME 2 Jupyter extension with:<br />' +\n","                        '<code>jupyter serverextension enable --py qiime2' +\n","                        ' --sys-prefix</code><br />then restart your server.' +\n","                        '<br /><br />(Interactive output not available on ' +\n","                        'static notebook viewer services like nbviewer.)';\n","    } else if (res.status === 409) {\n","        div.innerHTML = 'Visualization no longer in scope. Re-run this cell' +\n","                        ' to see the visualization.';\n","    } else if (res.ok) {\n","        url = res.url;\n","        div.innerHTML = '<iframe src=\\'' + url + '\\' style=\\'' +\n","                        'width: 100%; height: 700px; border: 0;\\'>' +\n","                        '</iframe><hr />Open in a: <a href=\\'' + url + '\\'' +\n","                        ' target=\\'_blank\\'>new window</a>'\n","    } else {\n","        div.innerHTML = 'Something has gone wrong. Check notebook server for' +\n","                        ' errors.';\n","    }\n","});\n","})(this.parentElement, '/qiime2/redirect?location=/tmp/qiime2-archive-j2p9_zwn')\" src=\"data:image/gif;base64,R0lGODlhAQABAIAAAP///wAAACH5BAEAAAAALAAAAAABAAEAAAICRAEAOw==\" /></div>"],"text/plain":["<visualization: Visualization uuid: 5eab02f5-6f18-48bc-b574-ed8ea89bba74>"]},"execution_count":8,"metadata":{},"output_type":"execute_result"}],"source":["# Visualize\n","q2.Visualization.load('cleanup/asv-table-nometaz.qzv')"]},{"cell_type":"markdown","id":"premier-whole","metadata":{"id":"premier-whole"},"source":["## Alpha Rarefaction Curves\n","### Checking the number of features per sample\n","In this step, the filtered ASV table will be used to view the alpha rarefaction curves.\n","\n","Alpha rarefaction curves can be used to visualize whether a sample has been sufficiently sequenced to represent its true diversity.\n","\n","To view the rarefaction curves for all the samples, check the number of features/OTUs per sample in the filtered ASV table visualization file. Run the code block below and go to the **Interactive Sample Detail** tab to check for the sample with the most numerous features/OTUs."]},{"cell_type":"code","execution_count":null,"id":"studied-belarus","metadata":{"id":"studied-belarus","outputId":"a24005d5-7c85-4c70-9ef3-c1a36e2b4d4f"},"outputs":[{"data":{"text/html":["<div><img onload=\"(function(div, url){\n","if (typeof require !== 'undefined') {\n","    var baseURL = require.toUrl('').split('/').slice(0, -2).join('/');\n","} else {\n","    var baseURL = JSON.parse(\n","        document.getElementById('jupyter-config-data').innerHTML\n","    ).baseUrl.slice(0, -1);\n","}\n","url = baseURL + url;\n","fetch(url).then(function(res) {\n","    if (res.status === 404) {\n","        div.innerHTML = 'Install QIIME 2 Jupyter extension with:<br />' +\n","                        '<code>jupyter serverextension enable --py qiime2' +\n","                        ' --sys-prefix</code><br />then restart your server.' +\n","                        '<br /><br />(Interactive output not available on ' +\n","                        'static notebook viewer services like nbviewer.)';\n","    } else if (res.status === 409) {\n","        div.innerHTML = 'Visualization no longer in scope. Re-run this cell' +\n","                        ' to see the visualization.';\n","    } else if (res.ok) {\n","        url = res.url;\n","        div.innerHTML = '<iframe src=\\'' + url + '\\' style=\\'' +\n","                        'width: 100%; height: 700px; border: 0;\\'>' +\n","                        '</iframe><hr />Open in a: <a href=\\'' + url + '\\'' +\n","                        ' target=\\'_blank\\'>new window</a>'\n","    } else {\n","        div.innerHTML = 'Something has gone wrong. Check notebook server for' +\n","                        ' errors.';\n","    }\n","});\n","})(this.parentElement, '/qiime2/redirect?location=/tmp/qiime2-archive-50x6qay3')\" src=\"data:image/gif;base64,R0lGODlhAQABAIAAAP///wAAACH5BAEAAAAALAAAAAABAAEAAAICRAEAOw==\" /></div>"],"text/plain":["<visualization: Visualization uuid: 5eab02f5-6f18-48bc-b574-ed8ea89bba74>"]},"execution_count":6,"metadata":{},"output_type":"execute_result"}],"source":["q2.Visualization.load('cleanup/asv-table-nometaz.qzv')"]},{"cell_type":"markdown","id":"removed-savings","metadata":{"id":"removed-savings"},"source":["### Making alpha rarefaction curves\n","In the **Interactive Sample Detail**, take note of the highest ASV count. To view the rarefaction curves for all the samples, the highest number of ASV count will be set as the `--p-max-depth`."]},{"cell_type":"code","execution_count":null,"id":"experimental-batch","metadata":{"scrolled":true,"id":"experimental-batch"},"outputs":[],"source":["!qiime diversity alpha-rarefaction \\\n","  --i-table cleanup/asv-table-nometaz.qza \\\n","  --p-max-depth 2000 \\\n","  --p-metrics 'shannon' \\\n","  --m-metadata-file metadata.tsv \\\n","  --o-visualization cleanup/asv-table-nometaz-arare.qzv \\\n","\n","# Visualize\n","q2.Visualization.load('cleanup/asv-table-nometaz-arare.qzv')"]},{"cell_type":"code","execution_count":null,"id":"fitting-mobility","metadata":{"id":"fitting-mobility","outputId":"0f034b75-daeb-4600-bfb5-43b5e1bc14e0"},"outputs":[{"data":{"text/html":["<div><img onload=\"(function(div, url){\n","if (typeof require !== 'undefined') {\n","    var baseURL = require.toUrl('').split('/').slice(0, -2).join('/');\n","} else {\n","    var baseURL = JSON.parse(\n","        document.getElementById('jupyter-config-data').innerHTML\n","    ).baseUrl.slice(0, -1);\n","}\n","url = baseURL + url;\n","fetch(url).then(function(res) {\n","    if (res.status === 404) {\n","        div.innerHTML = 'Install QIIME 2 Jupyter extension with:<br />' +\n","                        '<code>jupyter serverextension enable --py qiime2' +\n","                        ' --sys-prefix</code><br />then restart your server.' +\n","                        '<br /><br />(Interactive output not available on ' +\n","                        'static notebook viewer services like nbviewer.)';\n","    } else if (res.status === 409) {\n","        div.innerHTML = 'Visualization no longer in scope. Re-run this cell' +\n","                        ' to see the visualization.';\n","    } else if (res.ok) {\n","        url = res.url;\n","        div.innerHTML = '<iframe src=\\'' + url + '\\' style=\\'' +\n","                        'width: 100%; height: 700px; border: 0;\\'>' +\n","                        '</iframe><hr />Open in a: <a href=\\'' + url + '\\'' +\n","                        ' target=\\'_blank\\'>new window</a>'\n","    } else {\n","        div.innerHTML = 'Something has gone wrong. Check notebook server for' +\n","                        ' errors.';\n","    }\n","});\n","})(this.parentElement, '/qiime2/redirect?location=/tmp/qiime2-archive-cbp1nr6a')\" src=\"data:image/gif;base64,R0lGODlhAQABAIAAAP///wAAACH5BAEAAAAALAAAAAABAAEAAAICRAEAOw==\" /></div>"],"text/plain":["<visualization: Visualization uuid: ecd55c77-7753-42b6-b7eb-b0326e23f8a7>"]},"execution_count":5,"metadata":{},"output_type":"execute_result"}],"source":["# Visualize\n","q2.Visualization.load('cleanup/asv-table-nometaz-arare.qzv')"]},{"cell_type":"markdown","id":"hydraulic-penny","metadata":{"id":"hydraulic-penny"},"source":["### Choosing a sampling depth and making a phylogenetic tree\n","After seeing the rarefaction curves, we will select the sampling depth where we will rarefy the samples. Rarefaction will standardize the number of ASVs to the smallest number or ASVs in a sample which will allow for comparision between sites.\n","\n","NOTE: The smallest number of ASVs in a sample is not always used as the sampling depth. This decision should be based on the rarefaction curves. If the selected sampling depth is not on the plateau-ed part of the curve, problems in statistical analyses may arise as the actual diversity for other samples may be reduced.\n","\n","Before rarefying the samples and compute for alpha and beta diversity indices, run the code block below to make a tree file which will be used for phylogeny-based diversity analyses (e.g., Faith's PD and Unifrac)."]},{"cell_type":"code","execution_count":null,"id":"favorite-tournament","metadata":{"id":"favorite-tournament"},"outputs":[],"source":["# Generate a tree for phylogenetic diversity analyses\n","%mkdir cleanup/phylogeny\n","\n","!qiime phylogeny align-to-tree-mafft-fasttree \\\n","  --i-sequences cleanup/asv-rep-seqs-nometaz-250.qza \\\n","  --o-alignment cleanup/phylogeny/aligned-rep-seqs.qza \\\n","  --o-masked-alignment cleanup/phylogeny/masked-aligned-rep-seqs.qza \\\n","  --o-tree cleanup/phylogeny/unrooted-tree.qza \\\n","  --o-rooted-tree cleanup/phylogeny/rooted-tree.qza"]},{"cell_type":"markdown","id":"intimate-prevention","metadata":{"id":"intimate-prevention"},"source":["### Rarefying samples\n","As the rarefaction curves for the samples seem to have plateau-ed in the smallest number of ASVs (i.e., 927), sampling depth will be set to this number."]},{"cell_type":"code","execution_count":null,"id":"aboriginal-allah","metadata":{"id":"aboriginal-allah"},"outputs":[],"source":["# Alpha and Beta Diversity Analyses\n","!qiime diversity core-metrics-phylogenetic \\\n","  --i-phylogeny cleanup/phylogeny/rooted-tree.qza \\\n","  --i-table cleanup/asv-table-nometaz.qza \\\n","  --p-sampling-depth 801 \\\n","  --m-metadata-file metadata.tsv \\\n","  --output-dir cleanup/core-metrics-results"]},{"cell_type":"code","execution_count":null,"id":"accepted-compatibility","metadata":{"scrolled":true,"id":"accepted-compatibility"},"outputs":[],"source":["# Summarize ASV table (rarefied)\n","!qiime feature-table summarize \\\n","    --i-table cleanup/core-metrics-results/rarefied_table.qza \\\n","    --o-visualization cleanup/rarefied-table.qzv \\\n","    --m-sample-metadata-file metadata.tsv"]},{"cell_type":"code","execution_count":null,"id":"least-medium","metadata":{"id":"least-medium","outputId":"a9d29b43-51bf-4488-fa88-b65b62ce65a7"},"outputs":[{"data":{"text/html":["<div><img onload=\"(function(div, url){\n","if (typeof require !== 'undefined') {\n","    var baseURL = require.toUrl('').split('/').slice(0, -2).join('/');\n","} else {\n","    var baseURL = JSON.parse(\n","        document.getElementById('jupyter-config-data').innerHTML\n","    ).baseUrl.slice(0, -1);\n","}\n","url = baseURL + url;\n","fetch(url).then(function(res) {\n","    if (res.status === 404) {\n","        div.innerHTML = 'Install QIIME 2 Jupyter extension with:<br />' +\n","                        '<code>jupyter serverextension enable --py qiime2' +\n","                        ' --sys-prefix</code><br />then restart your server.' +\n","                        '<br /><br />(Interactive output not available on ' +\n","                        'static notebook viewer services like nbviewer.)';\n","    } else if (res.status === 409) {\n","        div.innerHTML = 'Visualization no longer in scope. Re-run this cell' +\n","                        ' to see the visualization.';\n","    } else if (res.ok) {\n","        url = res.url;\n","        div.innerHTML = '<iframe src=\\'' + url + '\\' style=\\'' +\n","                        'width: 100%; height: 700px; border: 0;\\'>' +\n","                        '</iframe><hr />Open in a: <a href=\\'' + url + '\\'' +\n","                        ' target=\\'_blank\\'>new window</a>'\n","    } else {\n","        div.innerHTML = 'Something has gone wrong. Check notebook server for' +\n","                        ' errors.';\n","    }\n","});\n","})(this.parentElement, '/qiime2/redirect?location=/tmp/qiime2-archive-jepa9e8m')\" src=\"data:image/gif;base64,R0lGODlhAQABAIAAAP///wAAACH5BAEAAAAALAAAAAABAAEAAAICRAEAOw==\" /></div>"],"text/plain":["<visualization: Visualization uuid: 804f4828-b69c-4bcb-82b6-123547b4a4c3>"]},"execution_count":4,"metadata":{},"output_type":"execute_result"}],"source":["# Visualize\n","q2.Visualization.load('cleanup/rarefied-table.qzv')"]},{"cell_type":"markdown","id":"taken-large","metadata":{"id":"taken-large"},"source":["## PCoA plots and statistical testing\n","### Visualizing beta diversity indices\n","After looking for the discontinuity of data using hierarchical clustering, grouping of sites will then be viewed in a multidimensional space.\n","\n","Run the code blocks below to view the PCoA plot of each beta diversity metric"]},{"cell_type":"code","execution_count":null,"id":"alpha-treasurer","metadata":{"id":"alpha-treasurer"},"outputs":[],"source":["# Jaccard distance\n","q2.Visualization.load('cleanup/core-metrics-results/jaccard_emperor.qzv')"]},{"cell_type":"code","execution_count":null,"id":"alternative-growth","metadata":{"id":"alternative-growth"},"outputs":[],"source":["# Bray-Curtis dissimilarity\n","q2.Visualization.load('cleanup/core-metrics-results/bray_curtis_emperor.qzv')"]},{"cell_type":"code","execution_count":null,"id":"different-momentum","metadata":{"id":"different-momentum"},"outputs":[],"source":["# Unweighted Unifrac\n","q2.Visualization.load('cleanup/core-metrics-results/unweighted_unifrac_emperor.qzv')"]},{"cell_type":"code","execution_count":null,"id":"solved-qualification","metadata":{"id":"solved-qualification","outputId":"b308190d-e0b6-4493-c506-45e37a8083f9"},"outputs":[{"data":{"text/html":["<div><img onload=\"(function(div, url){\n","if (typeof require !== 'undefined') {\n","    var baseURL = require.toUrl('').split('/').slice(0, -2).join('/');\n","} else {\n","    var baseURL = JSON.parse(\n","        document.getElementById('jupyter-config-data').innerHTML\n","    ).baseUrl.slice(0, -1);\n","}\n","url = baseURL + url;\n","fetch(url).then(function(res) {\n","    if (res.status === 404) {\n","        div.innerHTML = 'Install QIIME 2 Jupyter extension with:<br />' +\n","                        '<code>jupyter serverextension enable --py qiime2' +\n","                        ' --sys-prefix</code><br />then restart your server.' +\n","                        '<br /><br />(Interactive output not available on ' +\n","                        'static notebook viewer services like nbviewer.)';\n","    } else if (res.status === 409) {\n","        div.innerHTML = 'Visualization no longer in scope. Re-run this cell' +\n","                        ' to see the visualization.';\n","    } else if (res.ok) {\n","        url = res.url;\n","        div.innerHTML = '<iframe src=\\'' + url + '\\' style=\\'' +\n","                        'width: 100%; height: 700px; border: 0;\\'>' +\n","                        '</iframe><hr />Open in a: <a href=\\'' + url + '\\'' +\n","                        ' target=\\'_blank\\'>new window</a>'\n","    } else {\n","        div.innerHTML = 'Something has gone wrong. Check notebook server for' +\n","                        ' errors.';\n","    }\n","});\n","})(this.parentElement, '/qiime2/redirect?location=/tmp/qiime2-archive-gyupup0c')\" src=\"data:image/gif;base64,R0lGODlhAQABAIAAAP///wAAACH5BAEAAAAALAAAAAABAAEAAAICRAEAOw==\" /></div>"],"text/plain":["<visualization: Visualization uuid: 286834cf-9a42-4d5a-8aef-d9d18c4bfae4>"]},"execution_count":2,"metadata":{},"output_type":"execute_result"}],"source":["# Weighted Unifrac\n","q2.Visualization.load('cleanup/core-metrics-results/weighted_unifrac_emperor.qzv')"]},{"cell_type":"markdown","id":"downtown-worcester","metadata":{"id":"downtown-worcester"},"source":["### Alpha significance testing between/among groups\n","After visualizing in a multidimensional space and deciding on clusters, significance testing of the groups based on their alpha diversity will be done. This is explored to check if a cluster of sites is significantly more diverse than the other.\n","\n","Run the code blocks below to test for significance and view the results."]},{"cell_type":"code","execution_count":null,"id":"prompt-memphis","metadata":{"scrolled":true,"id":"prompt-memphis"},"outputs":[],"source":["# Make a directory for significance testing\n","%mkdir sig-test\n","\n","# Shannon\n","!qiime diversity alpha-group-significance \\\n","  --i-alpha-diversity cleanup/core-metrics-results/shannon_vector.qza \\\n","  --m-metadata-file metadata.tsv \\\n","  --o-visualization sig-test/shannon-group-significance.qzv\n","\n","q2.Visualization.load('sig-test/shannon-group-significance.qzv')"]},{"cell_type":"code","execution_count":null,"id":"distant-cocktail","metadata":{"id":"distant-cocktail"},"outputs":[],"source":["# Faith's PD\n","!qiime diversity alpha-group-significance \\\n","  --i-alpha-diversity cleanup/core-metrics-results/faith_pd_vector.qza \\\n","  --m-metadata-file metadata.tsv \\\n","  --o-visualization sig-test/faith_pd-group-significance.qzv\n","\n","q2.Visualization.load('sig-test/faith_pd-group-significance.qzv')"]},{"cell_type":"code","execution_count":null,"id":"touched-lindsay","metadata":{"id":"touched-lindsay","outputId":"cfa78955-cb1b-45c9-9979-55c7a75f2cae"},"outputs":[{"data":{"text/html":["<div><img onload=\"(function(div, url){\n","if (typeof require !== 'undefined') {\n","    var baseURL = require.toUrl('').split('/').slice(0, -2).join('/');\n","} else {\n","    var baseURL = JSON.parse(\n","        document.getElementById('jupyter-config-data').innerHTML\n","    ).baseUrl.slice(0, -1);\n","}\n","url = baseURL + url;\n","fetch(url).then(function(res) {\n","    if (res.status === 404) {\n","        div.innerHTML = 'Install QIIME 2 Jupyter extension with:<br />' +\n","                        '<code>jupyter serverextension enable --py qiime2' +\n","                        ' --sys-prefix</code><br />then restart your server.' +\n","                        '<br /><br />(Interactive output not available on ' +\n","                        'static notebook viewer services like nbviewer.)';\n","    } else if (res.status === 409) {\n","        div.innerHTML = 'Visualization no longer in scope. Re-run this cell' +\n","                        ' to see the visualization.';\n","    } else if (res.ok) {\n","        url = res.url;\n","        div.innerHTML = '<iframe src=\\'' + url + '\\' style=\\'' +\n","                        'width: 100%; height: 700px; border: 0;\\'>' +\n","                        '</iframe><hr />Open in a: <a href=\\'' + url + '\\'' +\n","                        ' target=\\'_blank\\'>new window</a>'\n","    } else {\n","        div.innerHTML = 'Something has gone wrong. Check notebook server for' +\n","                        ' errors.';\n","    }\n","});\n","})(this.parentElement, '/qiime2/redirect?location=/tmp/qiime2-archive-1gyzo25x')\" src=\"data:image/gif;base64,R0lGODlhAQABAIAAAP///wAAACH5BAEAAAAALAAAAAABAAEAAAICRAEAOw==\" /></div>"],"text/plain":["<visualization: Visualization uuid: 616317d4-d2bf-47aa-a709-262387b179c4>"]},"execution_count":8,"metadata":{},"output_type":"execute_result"}],"source":["q2.Visualization.load('sig-test/faith_pd-group-significance.qzv')"]},{"cell_type":"code","execution_count":null,"id":"cultural-microwave","metadata":{"id":"cultural-microwave"},"outputs":[],"source":["# Observed ASVs\n","!qiime diversity alpha-group-significance \\\n","  --i-alpha-diversity cleanup/core-metrics-results/observed_features_vector.qza \\\n","  --m-metadata-file metadata.tsv \\\n","  --o-visualization sig-test/observed_features-group-significance.qzv\n","\n","q2.Visualization.load('sig-test/observed_features-group-significance.qzv')"]},{"cell_type":"code","execution_count":null,"id":"balanced-policy","metadata":{"id":"balanced-policy","outputId":"bdaca6e7-03f1-444c-f7bd-043470171885"},"outputs":[{"data":{"text/html":["<div><img onload=\"(function(div, url){\n","if (typeof require !== 'undefined') {\n","    var baseURL = require.toUrl('').split('/').slice(0, -2).join('/');\n","} else {\n","    var baseURL = JSON.parse(\n","        document.getElementById('jupyter-config-data').innerHTML\n","    ).baseUrl.slice(0, -1);\n","}\n","url = baseURL + url;\n","fetch(url).then(function(res) {\n","    if (res.status === 404) {\n","        div.innerHTML = 'Install QIIME 2 Jupyter extension with:<br />' +\n","                        '<code>jupyter serverextension enable --py qiime2' +\n","                        ' --sys-prefix</code><br />then restart your server.' +\n","                        '<br /><br />(Interactive output not available on ' +\n","                        'static notebook viewer services like nbviewer.)';\n","    } else if (res.status === 409) {\n","        div.innerHTML = 'Visualization no longer in scope. Re-run this cell' +\n","                        ' to see the visualization.';\n","    } else if (res.ok) {\n","        url = res.url;\n","        div.innerHTML = '<iframe src=\\'' + url + '\\' style=\\'' +\n","                        'width: 100%; height: 700px; border: 0;\\'>' +\n","                        '</iframe><hr />Open in a: <a href=\\'' + url + '\\'' +\n","                        ' target=\\'_blank\\'>new window</a>'\n","    } else {\n","        div.innerHTML = 'Something has gone wrong. Check notebook server for' +\n","                        ' errors.';\n","    }\n","});\n","})(this.parentElement, '/qiime2/redirect?location=/tmp/qiime2-archive-ldxpse4s')\" src=\"data:image/gif;base64,R0lGODlhAQABAIAAAP///wAAACH5BAEAAAAALAAAAAABAAEAAAICRAEAOw==\" /></div>"],"text/plain":["<visualization: Visualization uuid: 39bbb358-e43d-4796-8a25-a428364864a9>"]},"execution_count":9,"metadata":{},"output_type":"execute_result"}],"source":["q2.Visualization.load('sig-test/observed_features-group-significance.qzv')"]},{"cell_type":"code","execution_count":null,"id":"spoken-virtue","metadata":{"id":"spoken-virtue"},"outputs":[],"source":["# Evenness\n","!qiime diversity alpha-group-significance \\\n","  --i-alpha-diversity cleanup/core-metrics-results/evenness_vector.qza \\\n","  --m-metadata-file metadata.tsv \\\n","  --o-visualization sig-test/evenness_vector-group-significance.qzv\n","\n","q2.Visualization.load('sig-test/evenness_vector-group-significance.qzv')"]},{"cell_type":"code","execution_count":null,"id":"toxic-exhibition","metadata":{"id":"toxic-exhibition","outputId":"85f1b8b0-0326-4c0a-b614-2fb5fff0a1dc"},"outputs":[{"data":{"text/html":["<div><img onload=\"(function(div, url){\n","if (typeof require !== 'undefined') {\n","    var baseURL = require.toUrl('').split('/').slice(0, -2).join('/');\n","} else {\n","    var baseURL = JSON.parse(\n","        document.getElementById('jupyter-config-data').innerHTML\n","    ).baseUrl.slice(0, -1);\n","}\n","url = baseURL + url;\n","fetch(url).then(function(res) {\n","    if (res.status === 404) {\n","        div.innerHTML = 'Install QIIME 2 Jupyter extension with:<br />' +\n","                        '<code>jupyter serverextension enable --py qiime2' +\n","                        ' --sys-prefix</code><br />then restart your server.' +\n","                        '<br /><br />(Interactive output not available on ' +\n","                        'static notebook viewer services like nbviewer.)';\n","    } else if (res.status === 409) {\n","        div.innerHTML = 'Visualization no longer in scope. Re-run this cell' +\n","                        ' to see the visualization.';\n","    } else if (res.ok) {\n","        url = res.url;\n","        div.innerHTML = '<iframe src=\\'' + url + '\\' style=\\'' +\n","                        'width: 100%; height: 700px; border: 0;\\'>' +\n","                        '</iframe><hr />Open in a: <a href=\\'' + url + '\\'' +\n","                        ' target=\\'_blank\\'>new window</a>'\n","    } else {\n","        div.innerHTML = 'Something has gone wrong. Check notebook server for' +\n","                        ' errors.';\n","    }\n","});\n","})(this.parentElement, '/qiime2/redirect?location=/tmp/qiime2-archive-jl61dcch')\" src=\"data:image/gif;base64,R0lGODlhAQABAIAAAP///wAAACH5BAEAAAAALAAAAAABAAEAAAICRAEAOw==\" /></div>"],"text/plain":["<visualization: Visualization uuid: f944adcd-0687-4095-90f2-4113cd917c6d>"]},"execution_count":10,"metadata":{},"output_type":"execute_result"}],"source":["q2.Visualization.load('sig-test/evenness_vector-group-significance.qzv')"]},{"cell_type":"markdown","id":"necessary-genesis","metadata":{"id":"necessary-genesis"},"source":["### Beta significance testing between/among groups\n","Then, significance testing of the groups based on their beta diversity will be examined. This is done to check if there is a significant difference in the composition of communities between/among groups.\n","\n","Run the code blocks below to test for significance and view the results."]},{"cell_type":"code","execution_count":null,"id":"enabling-ghana","metadata":{"scrolled":true,"id":"enabling-ghana"},"outputs":[],"source":["# Bray-Curtis\n","# Plastics-attached vs free-living\n","!qiime diversity beta-group-significance \\\n","  --i-distance-matrix cleanup/core-metrics-results/bray_curtis_distance_matrix.qza \\\n","  --m-metadata-file metadata.tsv \\\n","  --m-metadata-column environment1 \\\n","  --o-visualization sig-test/bray-curtis-environment1-significance.qzv \\\n","  --p-pairwise\n","\n","q2.Visualization.load('sig-test/bray-curtis-environment1-significance.qzv')"]},{"cell_type":"code","execution_count":null,"id":"respective-retirement","metadata":{"id":"respective-retirement","outputId":"a0023d41-e361-4a94-f1de-b923dced9f06"},"outputs":[{"name":"stdout","output_type":"stream","text":["\u001b[32mSaved Visualization to: sig-test/unweighted-unifrac-environment1-significance.qzv\u001b[0m\r\n"]},{"data":{"text/html":["<div><img onload=\"(function(div, url){\n","if (typeof require !== 'undefined') {\n","    var baseURL = require.toUrl('').split('/').slice(0, -2).join('/');\n","} else {\n","    var baseURL = JSON.parse(\n","        document.getElementById('jupyter-config-data').innerHTML\n","    ).baseUrl.slice(0, -1);\n","}\n","url = baseURL + url;\n","fetch(url).then(function(res) {\n","    if (res.status === 404) {\n","        div.innerHTML = 'Install QIIME 2 Jupyter extension with:<br />' +\n","                        '<code>jupyter serverextension enable --py qiime2' +\n","                        ' --sys-prefix</code><br />then restart your server.' +\n","                        '<br /><br />(Interactive output not available on ' +\n","                        'static notebook viewer services like nbviewer.)';\n","    } else if (res.status === 409) {\n","        div.innerHTML = 'Visualization no longer in scope. Re-run this cell' +\n","                        ' to see the visualization.';\n","    } else if (res.ok) {\n","        url = res.url;\n","        div.innerHTML = '<iframe src=\\'' + url + '\\' style=\\'' +\n","                        'width: 100%; height: 700px; border: 0;\\'>' +\n","                        '</iframe><hr />Open in a: <a href=\\'' + url + '\\'' +\n","                        ' target=\\'_blank\\'>new window</a>'\n","    } else {\n","        div.innerHTML = 'Something has gone wrong. Check notebook server for' +\n","                        ' errors.';\n","    }\n","});\n","})(this.parentElement, '/qiime2/redirect?location=/tmp/qiime2-archive-de67f151')\" src=\"data:image/gif;base64,R0lGODlhAQABAIAAAP///wAAACH5BAEAAAAALAAAAAABAAEAAAICRAEAOw==\" /></div>"],"text/plain":["<visualization: Visualization uuid: 10891af8-632b-4a8e-b3ab-9e55858e3dbf>"]},"execution_count":3,"metadata":{},"output_type":"execute_result"}],"source":["# Weighted Unifrac\n","# Plastics-attached vs free-living\n","!qiime diversity beta-group-significance \\\n","  --i-distance-matrix cleanup/core-metrics-results/unweighted_unifrac_distance_matrix.qza \\\n","  --m-metadata-file metadata.tsv \\\n","  --m-metadata-column environment1 \\\n","  --o-visualization sig-test/unweighted-unifrac-environment1-significance.qzv \\\n","  --p-pairwise\n","\n","q2.Visualization.load('sig-test/unweighted-unifrac-environment1-significance.qzv')"]},{"cell_type":"markdown","id":"expected-hypothesis","metadata":{"id":"expected-hypothesis"},"source":["### Exporting ASV tables\n","\n","We can export the ASV tables in a format we can use in other programs, such as R."]},{"cell_type":"code","execution_count":null,"id":"backed-actress","metadata":{"id":"backed-actress"},"outputs":[],"source":["# NOT RARIFIED\n","!qiime tools export --input-path cleanup/asv-table-nometaz.qza  --output-path exported-unrare \\\n","\n","!biom convert\\\n","    -i exported-unrare/feature-table.biom\\\n","    -o exported-unrare/feature-table-unrare.tsv\\\n","    --to-tsv"]},{"cell_type":"code","execution_count":null,"id":"public-perfume","metadata":{"id":"public-perfume"},"outputs":[],"source":["# RAREFIED\n","!qiime tools export --input-path cleanup/core-metrics-results/rarefied_table.qza  --output-path exported\n","!qiime tools export --input-path cleanup/asv-taxa-250.qza --output-path exported\n","\n","# Change the first line of biom-taxonomy.tsv (i.e. the header) to this:\n","# #ASVID taxonomy confidence\n","!sed '1c#ASVID\\ttaxonomy\\tconfidence' exported/taxonomy.tsv > exported/biom-taxonomy.tsv"]},{"cell_type":"code","execution_count":null,"id":"theoretical-numbers","metadata":{"id":"theoretical-numbers"},"outputs":[],"source":["!biom add-metadata \\\n","    -i exported/feature-table.biom \\\n","    -o exported/table-with-taxonomy.biom \\\n","    --observation-metadata-fp exported/biom-taxonomy.tsv \\\n","    --sc-separated taxonomy\n","\n","!biom convert\\\n","    -i exported/table-with-taxonomy.biom\\\n","    -o exported/feature-table-with-tax.tsv\\\n","    --to-tsv \\\n","    --header-key taxonomy"]},{"cell_type":"code","execution_count":null,"id":"packed-superior","metadata":{"id":"packed-superior","outputId":"863b9bd4-36b0-4a77-a442-866e0e6e9727"},"outputs":[{"name":"stdout","output_type":"stream","text":["\u001b[32mSaved FeatureTable[Frequency] to: cleanup/table-genus.qza\u001b[0m\n","\u001b[32mExported cleanup/table-genus.qza as BIOMV210DirFmt to directory exported-genus\u001b[0m\n"]}],"source":["# Genus level\n","!qiime taxa collapse \\\n","    --i-table cleanup/core-metrics-results/rarefied_table.qza \\\n","    --i-taxonomy cleanup/asv-taxa-250.qza \\\n","    --p-level 7 \\\n","    --o-collapsed-table cleanup/table-genus.qza\n","\n","!qiime tools export \\\n","    --input-path cleanup/table-genus.qza \\\n","    --output-path exported-genus\n","\n","!biom convert\\\n","    -i exported-genus/feature-table.biom\\\n","    -o exported-genus/feature-table-with-tax.tsv\\\n","    --to-tsv \\\n","    --header-key taxonomy"]},{"cell_type":"code","execution_count":null,"id":"incident-classification","metadata":{"id":"incident-classification"},"outputs":[],"source":["# Class level\n","!qiime taxa collapse \\\n","    --i-table cleanup/core-metrics-results/rarefied_table.qza \\\n","    --i-taxonomy cleanup/asv-taxa-250.qza \\\n","    --p-level 4 \\\n","    --o-collapsed-table cleanup/table-class.qza\n","\n","!qiime tools export \\\n","    --input-path cleanup/table-class.qza \\\n","    --output-path exported\n","\n","!biom convert\\\n","    -i exported/table-with-taxonomy.biom\\\n","    -o exported/feature-table-with-tax.tsv\\\n","    --to-tsv \\\n","    --header-key taxonomy"]},{"cell_type":"code","execution_count":null,"id":"narrow-pitch","metadata":{"id":"narrow-pitch","outputId":"29a6b678-bb85-4837-bade-5d5a1ec70f68"},"outputs":[{"name":"stdout","output_type":"stream","text":["\u001b[32mSaved FeatureTable[Frequency] to: cleanup/table-kingdom.qza\u001b[0m\n","\u001b[32mExported cleanup/table-kingdom.qza as BIOMV210DirFmt to directory exported-kingdom\u001b[0m\n"]}],"source":["# Kingdom level\n","!qiime taxa collapse \\\n","    --i-table cleanup/core-metrics-results/rarefied_table.qza \\\n","    --i-taxonomy cleanup/asv-taxa-250.qza \\\n","    --p-level 2 \\\n","    --o-collapsed-table cleanup/table-kingdom.qza\n","\n","!qiime tools export \\\n","    --input-path cleanup/table-kingdom.qza \\\n","    --output-path exported-kingdom\n","\n","!biom convert\\\n","    -i exported-kingdom/feature-table.biom\\\n","    -o exported-kingdom/feature-table-with-tax.tsv\\\n","    --to-tsv \\\n","    --header-key taxonomy"]},{"cell_type":"code","execution_count":null,"id":"monthly-western","metadata":{"id":"monthly-western"},"outputs":[],"source":["# Phylum level\n","!qiime taxa collapse \\\n","    --i-table cleanup/core-metrics-results/rarefied_table.qza \\\n","    --i-taxonomy cleanup/asv-taxa-250.qza \\\n","    --p-level 3 \\\n","    --o-collapsed-table cleanup/table-phylum.qza\n","\n","!qiime tools export \\\n","    --input-path cleanup/table-phylum.qza \\\n","    --output-path exported-phylum\n","\n","!biom convert\\\n","    -i exported-phylum/feature-table.biom\\\n","    -o exported-phylum/feature-table-with-tax.tsv\\\n","    --to-tsv \\\n","    --header-key taxonomy"]},{"cell_type":"markdown","id":"internal-plane","metadata":{"id":"internal-plane"},"source":["## ANCOM"]},{"cell_type":"code","execution_count":null,"id":"worth-sheep","metadata":{"id":"worth-sheep","outputId":"6e005fcc-b734-48e5-fbf7-2afbc68bb89a"},"outputs":[{"name":"stdout","output_type":"stream","text":["\u001b[32mSaved FeatureTable[Composition] to: cleanup/comp-table.qza\u001b[0m\r\n"]}],"source":["# Phylum level\n","!qiime composition add-pseudocount \\\n","    --i-table cleanup/table-phylum.qza \\\n","    --o-composition-table cleanup/comp-table.qza"]},{"cell_type":"code","execution_count":null,"id":"printable-conversion","metadata":{"id":"printable-conversion","outputId":"0a779c1a-8d3f-4b2f-da4c-a9158e09c20f"},"outputs":[{"name":"stdout","output_type":"stream","text":["\u001b[32mSaved Visualization to: cleanup/comp-ancom.qzv\u001b[0m\r\n"]},{"data":{"text/html":["<div><img onload=\"(function(div, url){\n","if (typeof require !== 'undefined') {\n","    var baseURL = require.toUrl('').split('/').slice(0, -2).join('/');\n","} else {\n","    var baseURL = JSON.parse(\n","        document.getElementById('jupyter-config-data').innerHTML\n","    ).baseUrl.slice(0, -1);\n","}\n","url = baseURL + url;\n","fetch(url).then(function(res) {\n","    if (res.status === 404) {\n","        div.innerHTML = 'Install QIIME 2 Jupyter extension with:<br />' +\n","                        '<code>jupyter serverextension enable --py qiime2' +\n","                        ' --sys-prefix</code><br />then restart your server.' +\n","                        '<br /><br />(Interactive output not available on ' +\n","                        'static notebook viewer services like nbviewer.)';\n","    } else if (res.status === 409) {\n","        div.innerHTML = 'Visualization no longer in scope. Re-run this cell' +\n","                        ' to see the visualization.';\n","    } else if (res.ok) {\n","        url = res.url;\n","        div.innerHTML = '<iframe src=\\'' + url + '\\' style=\\'' +\n","                        'width: 100%; height: 700px; border: 0;\\'>' +\n","                        '</iframe><hr />Open in a: <a href=\\'' + url + '\\'' +\n","                        ' target=\\'_blank\\'>new window</a>'\n","    } else {\n","        div.innerHTML = 'Something has gone wrong. Check notebook server for' +\n","                        ' errors.';\n","    }\n","});\n","})(this.parentElement, '/qiime2/redirect?location=/tmp/qiime2-archive-_mcwpjbg')\" src=\"data:image/gif;base64,R0lGODlhAQABAIAAAP///wAAACH5BAEAAAAALAAAAAABAAEAAAICRAEAOw==\" /></div>"],"text/plain":["<visualization: Visualization uuid: ae745913-09ad-4120-9270-b87a53df97aa>"]},"execution_count":5,"metadata":{},"output_type":"execute_result"}],"source":["!qiime composition ancom \\\n","    --i-table cleanup/comp-table.qza\\\n","    --m-metadata-file metadata.tsv \\\n","    --m-metadata-column environment1 \\\n","    --o-visualization cleanup/comp-ancom.qzv \\\n","\n","q2.Visualization.load('cleanup/comp-ancom.qzv')"]},{"cell_type":"code","execution_count":null,"id":"maritime-broadcasting","metadata":{"id":"maritime-broadcasting","outputId":"12540dea-b278-4eae-8c60-637fa5af651b"},"outputs":[{"data":{"text/html":["<div><img onload=\"(function(div, url){\n","if (typeof require !== 'undefined') {\n","    var baseURL = require.toUrl('').split('/').slice(0, -2).join('/');\n","} else {\n","    var baseURL = JSON.parse(\n","        document.getElementById('jupyter-config-data').innerHTML\n","    ).baseUrl.slice(0, -1);\n","}\n","url = baseURL + url;\n","fetch(url).then(function(res) {\n","    if (res.status === 404) {\n","        div.innerHTML = 'Install QIIME 2 Jupyter extension with:<br />' +\n","                        '<code>jupyter serverextension enable --py qiime2' +\n","                        ' --sys-prefix</code><br />then restart your server.' +\n","                        '<br /><br />(Interactive output not available on ' +\n","                        'static notebook viewer services like nbviewer.)';\n","    } else if (res.status === 409) {\n","        div.innerHTML = 'Visualization no longer in scope. Re-run this cell' +\n","                        ' to see the visualization.';\n","    } else if (res.ok) {\n","        url = res.url;\n","        div.innerHTML = '<iframe src=\\'' + url + '\\' style=\\'' +\n","                        'width: 100%; height: 700px; border: 0;\\'>' +\n","                        '</iframe><hr />Open in a: <a href=\\'' + url + '\\'' +\n","                        ' target=\\'_blank\\'>new window</a>'\n","    } else {\n","        div.innerHTML = 'Something has gone wrong. Check notebook server for' +\n","                        ' errors.';\n","    }\n","});\n","})(this.parentElement, '/qiime2/redirect?location=/tmp/qiime2-archive-_ndjp385')\" src=\"data:image/gif;base64,R0lGODlhAQABAIAAAP///wAAACH5BAEAAAAALAAAAAABAAEAAAICRAEAOw==\" /></div>"],"text/plain":["<visualization: Visualization uuid: 85edf951-0d7c-475c-9719-bee9f9f2fe3b>"]},"execution_count":2,"metadata":{},"output_type":"execute_result"}],"source":["q2.Visualization.load('cleanup/comp-ancom.qzv')"]}],"metadata":{"kernelspec":{"display_name":"Python 3","language":"python","name":"python3"},"language_info":{"codemirror_mode":{"name":"ipython","version":3},"file_extension":".py","mimetype":"text/x-python","name":"python","nbconvert_exporter":"python","pygments_lexer":"ipython3","version":"3.6.13"},"colab":{"name":"downstream-silva-138.1-v2.ipynb","provenance":[]}},"nbformat":4,"nbformat_minor":5}